Make EventId on WeekViewEvent generic for custom calendar fetches

udaysrinath opened the pull request:
Fixes issue : https://github.com/Quivr/Android-Week-View/issues/44

Note : 
(1) I rebased and that pulled in a lot of other changes from upstream, hope thats fine take a peek. 
(2) Tested for both Basic and Async Activity on the sample app and the change looks good.

@jhoobergs Please Review

Commit:
Make EventId on WeekViewEvent generic for custom calendar fetches

jhoobergs commented on library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java:
```diff
@@ -193,11 +216,6 @@ public boolean equals(Object o) {
 
     }
 
-    @Override
```
I think this method should remain. I guess you can return mId.hashCode() or something

jhoobergs commented on library/src/main/java/com/alamkanak/weekview/WeekView.java:
```diff
@@ -301,7 +301,7 @@ public boolean onSingleTapConfirmed(MotionEvent e) {
                 Collections.reverse(reversedEventRects);
                 for (EventRect eventRect : reversedEventRects) {
-                    if (eventRect.event.getId() != mNewEventId &&eventRect.rectF != null && e.getX() > eventRect.rectF.left && e.getX() < eventRect.rectF.right && e.getY() > eventRect.rectF.top && e.getY() < eventRect.rectF.bottom) {
+                    if (!eventRect.event.getId().equals(mNewEventId) && eventRect.rectF != null && e.getX() > eventRect.rectF.left && e.getX() < eventRect.rectF.right && e.getY() > eventRect.rectF.top && e.getY() < eventRect.rectF.bottom) {
```
I would write mNewEventId.equals(eventRect.event.getId()) to avoid possible null-pointer exceptions on the id of an event (mNewEventId won't be null)

jhoobergs commented on library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java:
```diff
@@ -113,10 +126,20 @@ public WeekViewEvent(long id, String name, String location, Calendar startTime,
+     * @param startTime The time when the event starts.
+     * @param endTime The time when the event ends.
+     */
+    public WeekViewEvent(int id, String name, Calendar startTime, Calendar endTime) {
```
Why did you add this one specifically ? I don't see the logic behind this.

udaysrinath replied:
These are overrides for the ```BaseActivity``` in the sample. I guess the helpers can move to the sample app, and dont need to be here.

jhoobergs reviewed (changes requested):
Some small comments.
Also: I think we should make it backward compatible, so don't change long return types to String, but create a new one for the string Id and let the old one use Integer.parse().

udaysrinath commented:
> Also: I think we should make it backward compatible, so don't change long return types to String, but create a new one for the string Id and let the old one use Integer.parse().

Done!

jhoobergs commented:
@udaysrinath 
The backward compatibility isn't complete yet: 
- The constructors with an int as id should all still work
- The xml attr with an int should still work (so create a seperate one for the new string attribute)

After this, the samples should still work without any changes. 
When that's the case, you can add extra examples that use a string as id

udaysrinath commented:
Good catch @jhoobergs! Hopefully I have caught all the backward compat scenarios. Also, updated the sample to accept both types. Added ```@Deprecated``` annotation explicitly.

jhoobergs commented on library/src/main/java/com/alamkanak/weekview/WeekView.java:
```diff
@@ -442,6 +443,7 @@ public WeekView(Context context, AttributeSet attrs, int defStyleAttr) {
             mNewEventColor = a.getColor(R.styleable.WeekView_newEventColor, mNewEventColor);
             mNewEventIconDrawable = a.getDrawable(R.styleable.WeekView_newEventIconResource);
             mNewEventId = a.getInt(R.styleable.WeekView_newEventId, mNewEventId);
+            mNewEventIdentifier =  (a.getString(R.styleable.WeekView_newEventId) != null)? a.getString(R.styleable.WeekView_newEventId) : mNewEventIdentifier;
```
I think you should also update the `mNewEventIdentifier`  when the int version is set. This can easily be solved when you use the setter `setNewEventId`. The attribute `int mNewEventId ` isn't needed anymore, you should just convert the int's to strings.

jhoobergs commented on library/src/main/java/com/alamkanak/weekview/WeekView.java:
```diff
@@ -1067,7 +1069,7 @@ private void drawEvents(Calendar date, float startFromPixel, Canvas canvas) {
                             topToUse = mHourHeight *  getPassedMinutesInDay(mMinTime, 0) / 60 + mCurrentOrigin.y + mHeaderHeight + mHeaderRowPadding * 2 + mHeaderMarginBottom + mTimeTextHeight/2 + mEventMarginVertical - marginTop;
 
-                        if(mEventRects.get(i).event.getId() != mNewEventId)
+                        if(mEventRects.get(i).event.getIdentifier() != mNewEventIdentifier)
```
String comparison should happen with `.equals()`

jhoobergs commented on library/src/main/java/com/alamkanak/weekview/WeekView.java:
```diff
@@ -1163,7 +1165,7 @@ private void drawEventTitle(WeekViewEvent event, RectF rect, Canvas canvas, floa
                 do {
                     // Ellipsize text to fit into event rect.
-                    if (event.getId() != mNewEventId)
+                    if (event.getIdentifier() != mNewEventIdentifier)
```
String comparison

jhoobergs commented on library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java:
```diff
@@ -65,13 +65,49 @@ public WeekViewEvent(long id, String name, int startYear, int startMonth, int st
+    @Deprecated
+    public WeekViewEvent(long id, String name, int startYear, int startMonth, int startDay, int startHour, int startMinute, int endYear, int endMonth, int endDay, int endHour, int endMinute) {
+        this.mId = Integer.toString((int) id);
+
```
you should call the constructor with the String attribute after converting instead of copy pasting.

jhoobergs replied:
The conversion should happen with `String.valueOf(id);`

jhoobergs commented on library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java:
```diff
@@ -88,9 +124,30 @@ public WeekViewEvent(long id, String name, String location, Calendar startTime,
+     */
+    @Deprecated
+    public WeekViewEvent(long id, String name, String location, Calendar startTime, Calendar endTime, boolean allDay, Shader shader) {
+        this.mId = Integer.toString((int) id);
```
Should also call the other constructor after conversion to string

jhoobergs commented on library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java:
```diff
@@ -101,22 +158,60 @@ public WeekViewEvent(long id, String name, String location, Calendar startTime,
-    public WeekViewEvent(long id, String name, String location, Calendar startTime, Calendar endTime) {
+    @Deprecated
+    public WeekViewEvent(long id, String name, String location, Calendar startTime, Calendar endTime, boolean allDay) {
+        this(Integer.toString((int) id), name, location, startTime, endTime, allDay, null);
```
This one is ok, but doesn't actually need to convert it to a string, the conversion would happen later the other constructor.

jhoobergs commented on library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java:
```diff
@@ -101,22 +158,60 @@ public WeekViewEvent(long id, String name, String location, Calendar startTime,
+     */
+    @Deprecated
+    public WeekViewEvent(long id, String name, Calendar startTime, Calendar endTime) {
+        this(Integer.toString((int) id), name, null, startTime, endTime);
```
Could also call `public WeekViewEvent(long id, String name, String location, Calendar startTime, Calendar endTime, boolean allDay)`, the conversion to int would happen later

jhoobergs commented on library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java:
```diff
@@ -174,14 +269,24 @@ public void setShader(Shader shader){
 
+    @Deprecated
+    public void setId(long id) {
+        this.mId = Integer.toString((int) id);
```
The conversion to int can cause problems: use `String.valueOf(id)`

jhoobergs reviewed (changes requested):
@udaysrinath Almost there 😄

udaysrinath commented:
@jhoobergs - Addressed the review comments, mind taking another look?

jhoobergs commented on library/src/main/java/com/alamkanak/weekview/WeekViewEvent.java:
```diff
@@ -174,14 +247,24 @@ public void setShader(Shader shader){
-    public void setId(long id) {
+    @Deprecated
+    public long getId() {
+        return Integer.parseInt(mId);
```
`Long.parseLong`

jhoobergs reviewed (changes requested):
Looks good.
Just one remark that I didn't see before

udaysrinath commented:
@jhoobergs - Updated PR and squashed commits into 1 commit. Can you merge and publish a new version, would like to consume it and ensure all is great! Thanks again for the review and your help.

jhoobergs commented:
@udaysrinath There is one thing I forgot (I'm sorry 😄 ) Can you also update the readme ? 
I'll test the code now

Commit:
Update Readme.md

udaysrinath commented:
Updated Readme, new version ```2.0.1```

udaysrinath commented:
@jhoobergs -Just noticed this, previously the gradle dependency was ``` compile 'com.github.quivr:android-week-view:2.0.2' ```, even though the readme was ```2.0.0``` can you explain that? Is there a discrepancy here? Is the usage section a bug that indicated ```2.0.2```?

jhoobergs commented:
@udaysrinath The changelog wasn't up to date, some bugfixes.
I pushed a new version called 2.2.0
Important: Automatic prefetching is disabled now, as described in the changelog

udaysrinath commented:
Hmm, cant seem to find anything with ```2.2.0``` if I add that as a dependency.

jhoobergs commented:
@udaysrinath Sometimes it takes some time. You can check http://search.maven.org/#search%7Cgav%7C1%7Cg%3A%22com.github.quivr%22%20AND%20a%3A%22android-week-view%22 to see if it should be available.
